### PR TITLE
Added login via LDAP

### PR DESCRIPTION
**Pull request by HelioStrike**
Added login via LDAP. The user can click on the 'Sign In' button, to reach a (not so shiny)login page.

Screenshot from 2019-05-25 09-12-22
 On entering the correct credentials, the user is logged into the Atlas, and can do Atlas stuff.

Screenshot from 2019-05-24 09-52-42

Wrong credentials give us an error message (I think I'll focus on the looks of the message in a future PR).
Screenshot from 2019-05-25 09-06-21

### CODE REVIEW

**Comment by cintiadr** [comment]
I know you copied this code from ID, but I think it's unnecessary complex to look for a user by username. 

I'd prefer if we refactor `searchUser` to be more like `searchGroups`, and delete the `convertUser` and  `searchRaw` methods. It's not really clear to me why the code is even like that in ID.

**Comment by cintiadr** [comment]
Probably you do not even want to load the user groups. Best case scenario you can add later to test if this user has permission to login (or if it's admin) based if it belongs to `atlas-user` or `atlas-admin`. But you have no reason to load all groups of a given user. 

It might be simpler if for now you just ignore it completely.

**Comment by HelioStrike** [comment]
@cintiadr I've updated the code. Please take a look at it again. :)

File: ldap.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ */
+exports.authenticate = (username, pass, cb) => {
+  log.debug(`${username}: will authenticate`);
+  // client used for authenticating users specially
+  const userClient = ldap.createClient({
+    url: url,
+  });
+  const userdn = `${userAttr.rdn}=${username},${userAttr.baseDn}`;
+  console.log(userdn);
+  console.log(pass);
```
Comment (bmamlin) [inline ldap.js]: Passwords should never be logged.

File: routes/authentication.js [hunk 2]
```diff
@@ -1,12 +1,46 @@
...
+        res.render('login', {user: req.session.user});
     });
 
+    //Validate login credentials and redirect accordingly
+    router.post('/login', function(req, res) {
+        ldapUtils.authenticate(req.body.username, req.body.password, function(err) {
+            //If authentication fails, redirect to login
+            if(err) {
+                console.log(err);
+                res.redirect('/login');
```
Comment (bmamlin) [inline routes/authentication.js]: Will this pass the error back to the login page so it can be displayed?

File: ldap.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Get a user's info from LDAP server based on the username
+ * @param  {string}   username  User's username
+ * @param  {Function} cb        cb(err, user)
+ */
+exports.getUser = (username, cb) => {
+  const base = `uid=${username},${userAttr.baseDn}`;
+  const options = {
+    scope: 'base',
+    attributes: ["uid","cn","sn","displayName","mail","userPassword"],
```
Comment (bmamlin) [inline ldap.js]: Is `userPassword` necessary here?

File: routes/authentication.js [hunk 4]
```diff
@@ -1,12 +1,46 @@
...
+            }
+            else {
+                //Fetch user details
+                ldapUtils.getUser(req.body.username, function(err, user) {
+                    console.log(user);
+                    //If unable to fetch user details, redirect to login
+                    if(err) {
+                        console.log(err);
+                        res.redirect('/login');
```
Comment (bmamlin) [inline routes/authentication.js]: Will this pass the error back to the login page so it can be displayed?

**Comment by HelioStrike** [comment]
@bmamlin I was just being polite to the more determined hackers!

Sorry, I missed that log, but the code's updated now! The error message bit wasn't implemented earlier, but I added it to the code, and you can find a screenshot in the first comment of the PR. :)

File: conf.example.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      "lastname": "sn",
+      "displayname": "displayName",
+      "email": "mail",
+      "password": "userPassword",
+      "secondaryemail": "otherMailbox",
+      "defaultObjectClass": [
+        "inetOrgPerson",
+        "extensibleObject"
+      ],
+      "usernameRegex": /^(?!.*(?:openmrs|adm))[a-zA-Z][a-zA-Z0-9.]{2,20}$/,
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply conf.example.js]: I pushed the change last night, I believe you are looking at an outdated version of the file. (I'm sort of new to PRs in github so I don't know how thats possible, but it says "Outdated" near the file name, so I assume that's the case)

File: ldap.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+});
+
+/* Private helper functions */
+
+// get the raw LDAP entry
+const searchRaw = (username, attributes, cb) => {
+  if (!_.isArray(attributes)) {
+    attributes = [attributes];
+  }
+  const getAll = username === '*';
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply ldap.js]: Same here, the file seems to be outdated.

File: conf.example.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      "rdn": "uid",
+
+      // corresponds with form input names
+      "username": "uid",
+      "firstname": "cn",
+      "lastname": "sn",
+      "displayname": "displayName",
+      "email": "mail",
+      "password": "userPassword",
+      "secondaryemail": "otherMailbox",
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply conf.example.js]: Here too

File: routes/authentication.js [hunk 8]
```diff
@@ -1,12 +1,46 @@
 var express = require('express');
+var ldap = require('ldapjs');
 var router = express.Router();
+var ldapUtils = require('../ldap');
+
+var client = ldap.createClient({
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply routes/authentication.js]: Nice catch! I just pushed the changes :)
Looks like ldapjs wasn't being used at all in this file, so I removed that line too.

File: routes/authentication.js (hunk 4 above)
Comment (HelioStrike) [reply routes/authentication.js]: I'm storing the error in the session, so that it can be displayed later. What do you think of it?

**Comment by cintiadr** [comment]
@bmamlin / @harsha89 , are you ok merging this PR? I'd like it merged so I can try to get staging back working, even if there's work to do :)

**Pull request merged**